src/gss_serialize.c

```diff
@@ -125,57 +198,48 @@ static int export_name(struct export_state *state,
 {
     int ret;
 
+    memset(exp_name, 0, sizeof(struct export_name));
+
     switch (name->type) {
     case GSSNTLM_NAME_NULL:
-        memset(exp_name, 0, sizeof(struct export_name));
-        return 0;
+        break;
     case GSSNTLM_NAME_ANON:
-        memset(exp_name, 0, sizeof(struct export_name));
         exp_name->type = EXP_NAME_ANON;
-        return 0;
+        break;
     case GSSNTLM_NAME_USER:
         exp_name->type = EXP_NAME_USER;
         if (name->data.user.domain) {
             ret = export_data_buffer(state, name->data.user.domain,
-                                     strlen(name->data.user.domain) + 1,
+                                     strlen(name->data.user.domain),
```

Comment:
I don't think this change is right.  `strlen()` doesn't include the NUL, so this creates data buffers that aren't terminated.   `import_data_buffer()` accidentally gets this right in the `alloc` case because `strndup()` will tack on another NUL.  However, the non-`alloc` case will be off-by-one in the `memcpy()`, and will produce something nonterminated.

It'd be a larger refactor, but I think it would simplify things to sliminate the `str` case entirely from `import_data_buffer()` and always NUL-terminate strings.

Reply:
> I don't think this change (and the others like it) is right. `strlen()` doesn't include the NUL, so this creates data buffers that aren't terminated.

This is intentional as the nul character is redundant, given we have the exact length for each buffer.

 `import_data_buffer()` accidentally gets this right in the `alloc` case because `strndup()` will tack on another NUL. 

this is not an accident, I checked it does it correctly.

However, the non-`alloc` case, in the `memcpy()`, will produce something nonterminated.

In all cases where a string is used we do allocate memory so it is not a problem in practice, but it is indeed an issue waiting to strike. I'll fix the import function to do the right thing for the 'str' case.

> 
> It'd be a larger refactor, but I think it would simplify things to eliminate the `str` case entirely from `import_data_buffer()` and always NUL-terminate strings.

I prefer to maintain this structure as compact as possible, so I'll just fix the code to cope.

Reply:
Actually , looking at it the way this works is that if you want to provide a pre-allocated buffer then you need to nul terminate it yourself.
I will fail if a pre-allocated buffer also has bool str set to True, you cannot ask us to terminate a buffer you control.